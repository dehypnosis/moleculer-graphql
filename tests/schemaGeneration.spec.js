import { ServiceBroker, Transporters } from "moleculer";
import { printSchema } from "graphql";
import { GraphQLGateway } from "..";
import authorSvc from "./types/Author";
import bookSvc from "./types/Book";
import chapterSvc from "./types/chapter";

jest.setTimeout(10000);

describe("Schema Generation", () => {
  describe("With A Single Broker", () => {
    // Globals for With a Single Broker
    let broker = null;
    let gateway = null;

    beforeAll(() => {
      broker = new ServiceBroker({
        nodeID: "gatewaySingle",
        namespace: "schemaGeneration"
      });

      broker.createService(authorSvc);
      broker.createService(bookSvc);
      broker.createService(chapterSvc);

      broker.start();

      gateway = new GraphQLGateway(broker);

      return gateway.start();
    });

    afterAll(() => gateway.stop);

    test("Should generate a consistant schema", () => {
      expect(printSchema(gateway.schema)).toMatchSnapshot();
    });
  });

  describe("With Multiple Brokers", () => {
    // Globals for With a Single Broker
    let broker = null;
    let authorBroker = null;
    let bookBroker = null;
    let chapterBroker = null;
    let gateway = null;

    beforeAll(async () => {
      broker = new ServiceBroker({
        nodeID: "gatewayMultiple",
        namespace: "schemaGeneration",
        transporter: new Transporters.MQTT("mqtt://localhost:1883")
      });

      authorBroker = new ServiceBroker({
        nodeID: "author",
        namespace: "schemaGeneration",
        transporter: new Transporters.MQTT("mqtt://localhost:1883")
      });

      bookBroker = new ServiceBroker({
        nodeID: "book",
        namespace: "schemaGeneration",
        transporter: new Transporters.MQTT("mqtt://localhost:1883")
      });

      chapterBroker = new ServiceBroker({
        nodeID: "chapter",
        namespace: "schemaGeneration",
        transporter: new Transporters.MQTT("mqtt://localhost:1883")
      });

      broker.start();

      gateway = new GraphQLGateway(broker);

      authorBroker.createService(authorSvc);
      bookBroker.createService(bookSvc);
      chapterBroker.createService(chapterSvc);

      await authorBroker.start();
      await bookBroker.start();
      await chapterBroker.start();
      await gateway.start();
    });

    afterAll(() =>
      Promise.all([
        gateway.stop(),
        authorBroker.stop(),
        bookBroker.stop(),
        chapterBroker.stop()
      ]));

    test("Should generate a consistant schema", () => {
      expect(printSchema(gateway.schema)).toMatchSnapshot();
    });
  });
});
