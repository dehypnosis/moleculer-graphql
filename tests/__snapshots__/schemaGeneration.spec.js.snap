// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema Generation With A Single Broker Should generate a consistant schema 1`] = `
"type Author {
  id: Int
  name: String
  books: [Book]
}

type Book {
  id: Int
  title: String
  authorId: Int
  year: Int
  author: Author
  chapters: [Chapter]
}

type Chapter {
  id: Int
  title: String
  bookId: Int
  book: Book
}

type Mutation {
  updateAuthor(input: UpdateAuthorInput!): UpdateAuthorPayload
  updateBook(input: UpdateBookInput!): UpdateBookPayload
}

type Query {
  author(id: Int!): Author
  authorOf(bookId: Int!): Author
  authors: [Author]
  book(id: Int!): Book
  books: [Book]
  booksByAuthor(authorId: Int!): [Book]
  chapter(id: Int!): Chapter
  chapters: [Chapter]
  chaptersInBook(bookId: Int!): [Chapter]
}

input UpdateAuthorInput {
  id: Int!
  clientMutationId: Int!
  name: String
}

type UpdateAuthorPayload {
  author: Author
  clientMutationId: Int
}

input UpdateBookInput {
  id: Int!
  clientMutationId: Int!
  title: String
}

type UpdateBookPayload {
  book: Book
  clientMutationId: Int
}
"
`;

exports[`Schema Generation With Multiple Brokers Should generate a consistant schema 1`] = `
"type Author {
  id: Int
  name: String
  books: [Book]
}

type Book {
  id: Int
  title: String
  authorId: Int
  year: Int
  author: Author
  chapters: [Chapter]
}

type Chapter {
  id: Int
  title: String
  bookId: Int
  book: Book
}

type Mutation {
  updateAuthor(input: UpdateAuthorInput!): UpdateAuthorPayload
  updateBook(input: UpdateBookInput!): UpdateBookPayload
}

type Query {
  author(id: Int!): Author
  authorOf(bookId: Int!): Author
  authors: [Author]
  book(id: Int!): Book
  books: [Book]
  booksByAuthor(authorId: Int!): [Book]
  chapter(id: Int!): Chapter
  chapters: [Chapter]
  chaptersInBook(bookId: Int!): [Chapter]
}

input UpdateAuthorInput {
  id: Int!
  clientMutationId: Int!
  name: String
}

type UpdateAuthorPayload {
  author: Author
  clientMutationId: Int
}

input UpdateBookInput {
  id: Int!
  clientMutationId: Int!
  title: String
}

type UpdateBookPayload {
  book: Book
  clientMutationId: Int
}
"
`;
